Libraries/LibGfx/VulkanContext.cpp

```diff
@@ -1,79 +1,152 @@
 /*
  * Copyright (c) 2024, Aliaksandr Kalenik <kalenik.aliaksandr@gmail.com>
+ * Copyright (c) 2025, Undefine <undefine@undefine.pl>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #include <AK/Format.h>
+#include <AK/NonnullOwnPtr.h>
 #include <AK/Vector.h>
 #include <LibGfx/VulkanContext.h>
 
 namespace Gfx {
 
-static ErrorOr<VkInstance> create_instance(uint32_t api_version)
+ErrorOr<NonnullRefPtr<VulkanContext>> VulkanContext::create()
 {
-    VkInstance instance;
+    auto context = adopt_ref(*new VulkanContext());
 
+    TRY(context->create_instance());
+    TRY(context->pick_physical_device());
+    TRY(context->create_logical_device_and_queue());
+
+#ifdef USE_VULKAN_IMAGES
+    TRY(context->create_command_pool());
+    TRY(context->allocate_command_buffer());
+    TRY(context->get_extensions());
+#endif
+
+    return context;
+}
+
+VulkanContext::VulkanContext()
+{
+    // NOTE: v1.1 needed for vkGetPhysicalDeviceFormatProperties2
+    // TODO: Try v1.0 if v1.1 is not supported as it's only needed for Vulkan images
+    m_api_version = VK_API_VERSION_1_1;
+}
+
+VulkanContext::~VulkanContext()
+{
+    if (m_logical_device != VK_NULL_HANDLE) {
+        vkDestroyDevice(m_logical_device, nullptr);
+    }
+
+    if (m_debug_messenger != VK_NULL_HANDLE) {
+        auto pfn_destroy_debug_messenger = reinterpret_cast<PFN_vkDestroyDebugUtilsMessengerEXT>(vkGetInstanceProcAddr(m_instance, "vkDestroyDebugUtilsMessengerEXT"));
+        pfn_destroy_debug_messenger(m_instance, m_debug_messenger, nullptr);
+    }
+
+    if (m_instance != VK_NULL_HANDLE) {
+        vkDestroyInstance(m_instance, nullptr);
+    }
+}
+
+ErrorOr<void> VulkanContext::create_instance()
+{
     VkApplicationInfo app_info {};
     app_info.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO;
     app_info.pApplicationName = "Ladybird";
     app_info.applicationVersion = VK_MAKE_VERSION(1, 0, 0);
     app_info.pEngineName = nullptr;
     app_info.engineVersion = VK_MAKE_VERSION(1, 0, 0);
-    app_info.apiVersion = api_version;
+    app_info.apiVersion = m_api_version;
+
+#ifdef VULKAN_DEBUG
```

Comment:
all the "ifdef VULKAN_DEBUG" are in other parts of Ladybird as "if constexpr (VULKAN_DEBUG) {" blocks. In the cpp it should be easy enough, but in the header might be a problem. Not sure what to suggest for this. Just wondering if you thought of this?

Reply:
In some places it is actually `#if`. I don't it matters that much in the end as both achieve the same and I think consistently using `#if` everywhere is better than using `if constexpr` in some places only as in some places it's not possible to use it.